sql/core/src/main/java/org/apache/spark/sql/execution/datasources/orc/JavaOrcColumnarBatchReader.java

```diff
@@ -0,0 +1,503 @@
+              if (fromColumn.isNull[index]) {
+                toColumn.appendNull();
+              } else {
+                toColumn.appendLong(vector[index]);
+              }
+              index += 1;
+            }
+          } else if (type instanceof TimestampType) {
+            TimestampColumnVector vector = ((TimestampColumnVector)fromColumn);
+            int index = 0;
+            while (index < batchSize) {
+              if (fromColumn.isNull[index]) {
+                toColumn.appendNull();
+              } else {
+                toColumn.appendLong(fromTimestampColumnVector(vector, index));
+              }
+              index += 1;
+            }
+          } else if (type instanceof FloatType) {
+            double[] vector = ((DoubleColumnVector)fromColumn).vector;
+            int index = 0;
+            while (index < batchSize) {
+              if (fromColumn.isNull[index]) {
+                toColumn.appendNull();
+              } else {
+                toColumn.appendFloat((float)vector[index]);
+              }
+              index += 1;
+            }
+          } else if (type instanceof DoubleType) {
+            double[] vector = ((DoubleColumnVector)fromColumn).vector;
+            int index = 0;
+            while (index < batchSize) {
+              if (fromColumn.isNull[index]) {
+                toColumn.appendNull();
+              } else {
+                toColumn.appendDouble(vector[index]);
+              }
+              index += 1;
+            }
+          } else if (type instanceof StringType || type instanceof BinaryType) {
+            BytesColumnVector vector = (BytesColumnVector)fromColumn;
+            int index = 0;
+            while (index < batchSize) {
+              if (fromColumn.isNull[index]) {
+                toColumn.appendNull();
+              } else {
+                toColumn.appendByteArray(
+                  vector.vector[index], vector.start[index], vector.length[index]);
+              }
+              index += 1;
+            }
+          } else if (type instanceof DecimalType) {
+            DecimalType decimalType = (DecimalType)type;
+            HiveDecimalWritable[] vector = ((DecimalColumnVector)fromColumn).vector;
+            int index = 0;
+            while (index < batchSize) {
+              if (fromColumn.isNull[index]) {
+                toColumn.appendNull();
+              } else {
+                appendDecimalWritable(
+                  toColumn,
+                  decimalType.precision(),
+                  decimalType.scale(),
+                  vector[index]);
+              }
+              index += 1;
+            }
+          } else {
+            throw new UnsupportedOperationException("Unsupported Data Type: " + type);
+          }
+        }
+      }
+      i += 1;
+    }
+    return true;
+  }
+
+  /**
+   * Default memory mode for ColumnarBatch.
+   */
+  public static final MemoryMode DEFAULT_MEMORY_MODE = MemoryMode.ON_HEAP;
```

Comment:
don't do this, pass a `memoryMode` parameter in `initBatch`, like parquet

Reply:
Then, we need the following, too.
```scala
        val vectorizedReader = new VectorizedParquetRecordReader(
          convertTz.orNull, enableOffHeapColumnVector && taskContext.isDefined)
```

```scala
  /**
   * The memory mode of the columnarBatch
   */
  private final MemoryMode MEMORY_MODE;

  public VectorizedParquetRecordReader(TimeZone convertTz, boolean useOffHeap) {
    this.convertTz = convertTz;
    MEMORY_MODE = useOffHeap ? MemoryMode.OFF_HEAP : MemoryMode.ON_HEAP;
  }
```

Reply:
I added like the above and didn't add `memoryMode` parameter at `initBatch`.